hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java

```diff
@@ -1329,8 +1328,10 @@ public static class EntityB {
 			return withRegisteredAssociationKeys(
 					() -> {
 						final DomainResult<?> keyResult;
-						if ( notFoundAction != null ) {
-							if ( sideNature == ForeignKeyDescriptor.Nature.KEY ) {
+						if ( sideNature == ForeignKeyDescriptor.Nature.KEY ) {
+							// If the key side is non-nullable we also need to add the keyResult
+							// to be able to manually check invalid foreign key references
+							if ( notFoundAction == NotFoundAction.EXCEPTION || !isInternalLoadNullable ) {
```

Comment:
If there is a not found action we always have to fetch the key result to do some sort of checking. I think the original if block was correct. Maybe you want to pair together on this?

Reply:
Restored this to always add the key result when there's a not found action.

Also updated the PR with reviewed tests and a new commit to handle checking when polymorphic associated entities should be fetched like we discussed, even in the case of `Embeddable`s.